lib/libc/gen/readdir.c

```diff
@@ -56,10 +56,10 @@ struct dirent *_readdir_unlocked(DIR *dirp, int skipdeleted) {
       dirp->dd_loc = 0;
     }
     if (dirp->dd_loc == 0 && !(dirp->dd_flags & __DTF_READALL)) {
-      dirp->dd_seek = lseek(dirp->dd_fd, (off_t)0, SEEK_CUR);
```

Comment:
So far `ls` command worked without an issue... why this change is needed?

Reply:
`ls` still works fine, but there is an issue with `find` - without this change, the last `readdir` call in this loop: https://git.suckless.org/sbase/file/find.c.html#l981 returns 0, but also sets errno to `EINVAL` and we get a lot of confusing `readdir: Invalid argument` messages after each recursive call to `find` function. The errno value in `readdir` is set by the last `lseek` call - after all the directory entries have been read by `getdents`, then the current file read offset is at the end of directory, so `lseek` sets  `EINVAL`.

Reply:
This requires further investigation. I think `find` was implemented under assumptions of `readdir` returns errors as Linux `glibc` would do. Obviously there may be some differences between `glibc` and `libc` from NetBSD.

Reply:
`readdir` in glibc does not return errno if the end of directory is reached unlinke NetBSD's `readdir`, thus suckless's `find` was probably implemented under such assumption. It seems, that we can either add a patch for `find`, so that no confusing warnings are printed, or we can adjust our `readdir` function, so that it does not return `errno==EINVAL` on end of directory.

Reply:
1. [readdir](https://pubs.opengroup.org/onlinepubs/9699919799/functions/readdir.html) as defined by POSIX standard does not allow the function to return `EINVAL`
2. Linux [readdir](https://man7.org/linux/man-pages/man3/readdir.3.html) only returns `EBADF`
3. `readdir` from *BSD says (in the manual page) that the function may also fail and set errno for any of the errors specified for the routine [getdents(2)](https://www.freebsd.org/cgi/man.cgi?query=getdents&sektion=2&apropos=0&manpath=FreeBSD+13.1-RELEASE+and+Ports)

IMO we should fix our `readdir` not to propagate unwanted errors from `getdents`. A user should never need to reach directly for `getdents` system call to perform directory reading. Please filter out `errno` on return path from `readdir` function. It must not return any other values than mentioned in POSIX standard.

Reply:
Please put `lseek` into original position, but clear `errno` afterwards.

Reply:
`lseek` is back in the original position and we don't return `errno` on end of directory: https://github.com/cahirwpz/mimiker/pull/1361/commits/0a28ff65ea3bd519f835bbbeb1a279b98113bc49